node/src/actors/chain_manager/mining.rs

```diff
@@ -314,6 +332,53 @@ impl ChainManager {
                     }
                 })
                 .flatten()
+                // Refrain from trying to resolve any more requests if we have already hit the limit
+                // of retrievals per epoch.
+                .and_then(move |vrf_proof| {
+                    let mut start_retrieval_count = cloned_retrieval_count.load(Relaxed);
+                    let mut final_retrieval_count = start_retrieval_count.saturating_add(added_retrieval_count);
+
+                    if final_retrieval_count > maximum_retrieval_count {
+                        log::info!("{} Refrained from resolving data request {} for epoch #{} because it contains {} \
```

Comment:
This line is very helpful but it may be too long for the logs. I was thinking about printing the full help only once and a short line afterwards, but that's just adding complexity for little benefit. Not sure what to do about that.